### PR TITLE
Martin/2277 dkg 1 to 1 msg

### PR DESCRIPTION
**Pull request by arrivets**
This PR addresses issue  [2277](https://github.com/dapperlabs/flow-go/issues/2277)  

It implements an engine that enables DKG nodes to send and receive private messages.

The engine itself implements the `DKGProcessor` interface of the crypto dkg package, although all the methods are stubbed-out apart from `PrivateSend`.

* Create engine to receive and send DKG messages
* Implement DKGProcessor interface (with stubs for all methods other
  than PrivateSend)
* Add test for DKG controller PrivateSend and Process
* Add tests for invalid DKG messages
* Move DKG controller to its own module
* Add Epoch field to DKGMessage

### CODE REVIEW

File: engine/consensus/dkg/processor/engine.go [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+	conduit     network.Conduit
+	msgCh       chan dkg.DKGMessage
+	committee   flow.IdentifierList
+	myIndex     int
+	epochID     flow.Identifier
+	epochIDLock sync.Mutex
+}
+
+// New returns a new DKGProcessor engine. Instances participanting in a common
+// DKG protocol should use the same _ordered_ list of committee identifiers,
```
Comment (jordanschalm) [inline engine/consensus/dkg/processor/engine.go:35]:
```suggestion
// New returns a new DKGProcessor engine. Instances participating in a common
// DKG protocol must use the same _ordered_ list of node identifiers,
```

These appear to be the standard node ID -- may be best to use that terminology here. "Committee identifiers" made me assume these were a different ID

File: engine/consensus/dkg/processor/engine.go [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+		return fmt.Errorf("wrong epoch id. Got %v, want %v", msg.EpochID, e.epochID)
+	}
+
+	// check that the message's origin is not out of range
+	if msg.Orig >= len(e.committee) || msg.Orig < 0 {
+		return fmt.Errorf("origin id out of range: %d", msg.Orig)
+	}
+
+	// check that the message's origin matches the sender's flow identifier
+	committeeID := e.committee[msg.Orig]
```
Comment (jordanschalm) [inline engine/consensus/dkg/processor/engine.go:149]:
```suggestion
	nodeID := e.committee[msg.Orig]
```

Again to match the convention used elsewhere in the codebase

File: engine/consensus/dkg/processor/engine.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+	default:
+		return fmt.Errorf("invalid event type (%T)", event)
+	}
+}
+
+func (e *Engine) onDKGMessage(originID flow.Identifier, msg dkg.DKGMessage) error {
+
+	// check that the message corresponds to the current epoch
+	e.epochIDLock.Lock()
+	defer e.epochIDLock.Unlock()
```
Comment (jordanschalm) [inline engine/consensus/dkg/processor/engine.go:138]: The `engine.Unit` has an embedded mutex that can be used for synchronizing the engine -- we can use that here. 
```suggestion
	e.unit.Lock()
	defer e.unit.Unlock()
```

File: engine/consensus/dkg/processor/engine.go [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+// because senders and recipients are identified by their index in this list.
+// The msgCh is used to forward incoming DKG messages to the consumer, such as
+// the Controller implemented in the DKG module.
+func New(
+	logger zerolog.Logger,
+	net module.Network,
+	me module.Local,
+	msgCh chan dkg.DKGMessage,
+	committee flow.IdentifierList,
+	epochID flow.Identifier) (*Engine, error) {
```
Comment (jordanschalm) [inline engine/consensus/dkg/processor/engine.go:45]: Epochs within a chain are identified by a `uint64` counter. Does this `epochID` need to be unique across different possible chains?

File: engine/consensus/dkg/processor/engine_test.go [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+			t.Fatal("Timeout waiting for dkg message")
+		default:
+			if received {
+				require.Equal(t, msg, receivedMessage)
+				return
+			}
+			time.Sleep(10 * time.Millisecond)
+		}
+	}
+}
```
Comment (jordanschalm) [inline engine/consensus/dkg/processor/engine_test.go:193]:
```suggestion
	var received dkg.DKGMessage
	// check that the message has been received and forwarded to the msgCh
	unittest.AssertReturnsBefore(func() {
		received = <-msgCh
	}, time.Second)
	require.Equal(t, msg, received)
}
```

File: engine/consensus/dkg/processor/engine.go (hunk 4 above)
Comment (arrivets) [reply engine/consensus/dkg/processor/engine.go]: Thanks! Was waiting for you to confirm the type :)
No I don't think they need to be unique across chains.

File: engine/consensus/dkg/processor/engine_test.go (hunk 5 above)
Comment (arrivets) [reply engine/consensus/dkg/processor/engine_test.go]: nice

File: engine/consensus/dkg/processor/engine.go [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+// ID.
+type Engine struct {
+	unit      *engine.Unit
+	log       zerolog.Logger
+	me        module.Local
+	conduit   network.Conduit
+	msgCh     chan dkg.DKGMessage
+	committee flow.IdentifierList
+	myIndex   int
+	epochID   uint64
```
Comment (jordanschalm) [inline engine/consensus/dkg/processor/engine.go:29]:
```suggestion
	epochCounter   uint64
```

Another minor tweak to match terminology we use elsewhere

**Review by jordanschalm — APPROVED** [review]
One last terminology comment. Otherwise LGTM - nice work!

File: engine/consensus/dkg/processor/engine.go [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+		return fmt.Errorf("invalid event type (%T)", event)
+	}
+}
+
+func (e *Engine) onDKGMessage(originID flow.Identifier, msg dkg.DKGMessage) error {
+
+	// check that the message corresponds to the current epoch
+	e.unit.Lock()
+	defer e.unit.Unlock()
+	if e.epochID != msg.EpochID {
```
Comment (tarakby) [inline engine/consensus/dkg/processor/engine.go:137]: 👌

File: engine/consensus/dkg/processor/engine_test.go [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+// message destination parameter (index in committee list) to the corresponding
+// public Identifier, and successfully sends a DKG message to the intended
+// recipient through the network conduit.
+func TestPrivateSend_Valid(t *testing.T) {
+
+	// initialize the list of dkg participants
+	committee := unittest.IdentifierListFixture(2)
+
+	// define epoch ID
+	epochID := rand.Uint64()
```
Comment (tarakby) [inline engine/consensus/dkg/processor/engine_test.go:37]: `math/rand` is always initialized by the same seed, so it's better to call `rand.Seed(time.Now().UnixNano())` once for each test to randomize the test.

Comment (tarakby) [reply engine/consensus/dkg/processor/engine_test.go:37]: Well it's not very important here as only the epoch ID is random which doesn't really impact the tests.

**Review by tarakby — APPROVED** [review]
Nice and clean 👌

**Pull request merged**